File: 0456-132-pattern/0456-132-pattern.cpp
class Solution {
public:
    bool find132pattern(vector<int>& nums) {
        int n = nums.size();
        vector<int> minElementToLeft(n,0);
        minElementToLeft[0]=nums[0];
        for(int i=1;i<n;i++) minElementToLeft[i] = min(nums[i],minElementToLeft[i-1]);
        //Stack to find NEXT GREATER ELEMENT TO LEFT
        stack<pair<int,int>> st;
        vector<int> NGLIndex(n,-1);
        for(int i=0;i<n;i++) {
            if(st.empty()) {
                st.push({nums[i],i});
            } else {
                while(!st.empty() && st.top().first<=nums[i]) {
                    st.pop();
                }
                if(!st.empty()) {
                    NGLIndex[i] = st.top().second;
                    
                } 
                st.push({nums[i],i});

            }
        }
        for(int k=n-1;k>0;k--) {
            int j = NGLIndex[k];
            if(j==0 || j==-1) continue;
            if(minElementToLeft[j-1]<nums[k]) return true;
        }
        return false;
    }
};

File: 0456-132-pattern/0456-132-pattern.py
class Solution:
    def find132pattern(self, nums):
        n = len(nums)
        minElementToLeft = [0] * n
        minElementToLeft[0] = nums[0]
        
        for i in range(1, n):
            minElementToLeft[i] = min(nums[i], minElementToLeft[i - 1])
        
        # Stack to find NEXT GREATER ELEMENT TO LEFT
        st = []
        NGLIndex = [-1] * n
        
        for i in range(n):
            while st and st[-1][0] <= nums[i]:
                st.pop()
            if st:
                NGLIndex[i] = st[-1][1]
            st.append((nums[i], i))
        
        for k in range(n - 1, 0, -1):
            j = NGLIndex[k]
            if j == 0 or j == -1:
                continue
            if minElementToLeft[j - 1] < nums[k]:
                return True
        
        return False
